### ISSUE TITLE
Repetitive lag when moving cursor

### ISSUE THREAD

**Issue by rlch** [nvim-neo-tree/neo-tree.nvim#268]
Hey, sorry to be the bearer of bad news again; but after switching to `main` to try out the fix for #224, I've been experiencing rather large delays at what seems like  ~2ish second intervals. 

Under the same conditions as stated in #224, I've attached a video demonstrating this issue:

https://user-images.githubusercontent.com/35163478/162868245-a5ee5277-a514-4a8d-8982-c27085e80e59.mov

**Comment by cseickel** [comment]
Does the issue go away when you disable diagnostics?

Does the issue happen with a minimal init? You can try out a minimal config by grabbing this file from the project: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/v2.x/tests/mininit.lua

Then open nvim with:
```
nvim --noplugin -u mininit.lua
```

**Comment by nyngwang** [comment]
To @cseickel, @rlch:

(**important note**: I'm not sure whether this is related. But this didn't happen before)

I didn't experience serious delay, but the fans of my laptop spun much faster(due to the temperature of `CPU PECI` getting higher). I suspect this is due to the process called `com.broadcom.mes.systemextension` occupying >100% CPU usage.

https://user-images.githubusercontent.com/24765272/163813600-52fe77bb-4f60-4aca-8b3b-6a462b875d07.mov

The process is probably related to the file-protection functionality installed by the company I'm working on. But I think this might provide some hint(s) for you to debug.

**Comment by soifou** [comment]
@cseickel I don't know either if this is related but I notice also some lags since few releases I'd like to share.

For me it appears when the sidebar is opened with sumneko_lua attached (in my nvim config).
The editor feels laggy when I'm entering insert mode, when I popup telescope, when I toggle some (but not all) nodes repeatedly. It becomes responsive again when I close the sidebar. 
When I reveal the sidebar again I notice a slight delay before it shows up.

I did not notice any lag with other LSP client nor project folder for the moment.

I dig a bit and found that particular commit [4f3feb4](https://github.com/nvim-neo-tree/neo-tree.nvim/commit/4f3feb48f96088aa045edd2d8fd32d4b98652b60) introduce this lag. More precisely [this line](https://github.com/nvim-neo-tree/neo-tree.nvim/commit/4f3feb48f96088aa045edd2d8fd32d4b98652b60#diff-b167bfc574a6e3298e7623806eba312a0bb9a24ee45cdcf30fb0b547c0a44282R259) seems the culprit.

As a workaround I have overriden the `modified` component to return nothing and the lag is gone:

```lua
            components = {
                modified = function(_, _, _) return {} end,
            },
```

I'll try to reproduce it with a minimal init... if I can ! 
And btw thanks so much for this great plugin :)

**Comment by cseickel** [comment]
Thanks for digging into this @soifou.  One thing I noticed what that line is that I am just swallowing up errors. It might give us a clue if you replace that `modified` function with one that echos out any errors:

```lua
M.modified = function(config, node, state)
  local success, bufnr = pcall(vim.fn.bufnr, node.path)
  if not success or bufnr < 0 then
    print("Error getting bufnr:", bufnr)
    return {}
  end
  local success2, modified = pcall(vim.api.nvim_buf_get_option, bufnr, "modified")
  if success2 and modified then
    return {
      text = " " .. (config.symbol or "[+]"),
      highlight = config.highlight or highlights.MODIFIED,
    }
  elseif not success2 then
    print("Error getting modified:", modified)
  else
    return {}
  end
end
```

**Comment by soifou** [comment]
Thanks, I modified a little your prints to get more verbose on the node path.

TLDR ;
- When I (re)open the sidebar; `modified`  called twice for each visible nodes
- When I open a node dir: normal behaviour, `modified` is called for each visible file nodes only one time
- When I open a file from the treeview : nothing
- When I modify/undo/save a buffer , `modified`  is called 3 times for each visible file nodes
- When I popup telescope ; 3 times for each visible file nodes.
- When I open a file from telescope : going crazy xD
- But this does not explain the lag I get with sumneko attached

--- 

On neo-tree root folder starting with `nvim -c "Neotree reveal toggle"` I get:

```
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Press ENTER or type command to continue
```

The sidebar reveal. Then I recheck `:messages` indicates same as above but twice:

```
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

**Steps below are message cleared on each iteration.**

Opening the node `lua/` seems to have correct behaviour

```
Error getting bufnr: -1 from node.path lua/neo-tree.lua
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

Opening telescope I get 3 duplicates

```
Error getting bufnr: -1 from node.path lua/neo-tree.lua
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Error getting bufnr: -1 from node.path lua/neo-tree.lua
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Error getting bufnr: -1 from node.path lua/neo-tree.lua
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

Opening lua/neo-tree.lua seems ok:

```
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

Going to insert mode/ undo change / save change:

```
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

Finally the most weird: opening another file (completion.lua here) in telescope while the sidebar is open

```
Success getting bufnr: 1500 from node.path lua/neo-tree/command/completion.lua
Success getting bufnr: 737 from node.path lua/neo-tree/command/init.lua
Success getting bufnr: 1477 from node.path lua/neo-tree/command/parser.lua
Success getting bufnr: 1552 from node.path lua/neo-tree/sources/common/commands.lua
Success getting bufnr: 116 from node.path lua/neo-tree/sources/common/components.lua
Success getting bufnr: 1622 from node.path lua/neo-tree/sources/common/container.lua
Success getting bufnr: 1119 from node.path lua/neo-tree/sources/common/file-items.lua
Success getting bufnr: 1659 from node.path lua/neo-tree/sources/common/file-nesting.lua
Success getting bufnr: 470 from node.path lua/neo-tree/sources/common/help.lua
Success getting bufnr: 1358 from node.path lua/neo-tree/sources/manager.lua
Success getting bufnr: 1304 from node.path lua/neo-tree/defaults.lua
Success getting bufnr: 1551 from node.path lua/neo-tree/log.lua
Success getting bufnr: 425 from node.path lua/neo-tree/utils.lua
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Success getting bufnr: 1500 from node.path lua/neo-tree/command/completion.lua
Success getting bufnr: 737 from node.path lua/neo-tree/command/init.lua
Success getting bufnr: 1477 from node.path lua/neo-tree/command/parser.lua
Success getting bufnr: 1304 from node.path lua/neo-tree/defaults.lua
Success getting bufnr: 1551 from node.path lua/neo-tree/log.lua
Success getting bufnr: 425 from node.path lua/neo-tree/utils.lua
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
Success getting bufnr: 1500 from node.path lua/neo-tree/command/completion.lua
Success getting bufnr: 737 from node.path lua/neo-tree/command/init.lua
Success getting bufnr: 1477 from node.path lua/neo-tree/command/parser.lua
Success getting bufnr: 1304 from node.path lua/neo-tree/defaults.lua
Success getting bufnr: 1551 from node.path lua/neo-tree/log.lua
Success getting bufnr: 425 from node.path lua/neo-tree/utils.lua
Success getting bufnr: 12 from node.path lua/neo-tree.lua
Success getting modified: true
Error getting bufnr: -1 from node.path CONTRIBUTING.md
Error getting bufnr: -1 from node.path LICENSE
Error getting bufnr: -1 from node.path Makefile
Error getting bufnr: -1 from node.path README.md
Error getting bufnr: -1 from node.path release.sh
```

**Comment by cseickel** [comment]
I had expected that all those calls would be super quick and it shouldn't have any impact on performance. If disabling that call made it faster though then I guess that assumption was wrong.

I think I can rewrite that so the modified check is cached for each file the first time and only updates in response to the event.

**Comment by soifou** [comment]
> the components get called on every render, which can happen a lot when diagnostics are updating

Thanks, you point me to the right direction! My sumneko workspace is pretty big since I require a lot of runtime files and by default it tries to diagnose the whole workspace. I significantly reduce the lag with the following settings in sumneko_lua:

```lua
            diagnostics = {
                enable = true,
                libraryFiles = "Opened",
                neededFileStatus = "Opened",
           },
```

> I think I can rewrite that so the modified check is cached for each file the first time and only updates in response to the event.

I think that would be a nice optimization or at least avoiding the trigger in your modified handler on some  unamed/readonly buffers such as `TelescopePrompt` or filetype like `PackerStatus`, `qf` ?

**Comment by cseickel** [comment]
> I think that would be a nice optimization or at least avoiding the trigger in your modified handler on some unamed/readonly buffers such as `TelescopePrompt` or filetype like `PackerStatus`, `qf` ?

That's a good point. I have a lot of triggers to refresh since there are so many status icons being shown, but I can certainly go in and filter out events related to these utility buffers.

**Comment by cseickel** [comment]
@soifou I just improved the modified marker handling. Can you try out `main` and tell me how it works?

**Comment by soifou** [comment]
@cseickel I have no more lags with theses changes, good job 👍🏼

**Comment by cseickel** [comment]
@rlch and anyone else watching this issue. I've now done a lot to increase efficiency and reduce unneeded refreshes, and I expect these issues to be resolved now. Please let me know if you are still experiencing any issues using the `main` branch.

If I don't get any bug reports, I'll release these changes tonight. I'll also close this issue at the same time unless @rlch reports that his particular issue is still happening.

**Comment by rlch** [comment]
Hey @cseickel, thanks so much for all your work on this; I'll try it out and let you know tomorrow.

**Comment by andersevenrud** [comment]
I've been running main and stock settings for a few days now (and did an update last night). Unless my problems lies elsewhere then I don't think this issue has been solved. Cursor/general responsiveness still starts to lag after a while of usage :(

**Comment by cseickel** [comment]
> I've been running main and stock settings for a few days now (and did an update last night). Unless my problems lies elsewhere then I don't think this issue has been solved. Cursor/general responsiveness still starts to lag after a while of usage :(

@andersevenrud Have you checked the process stats when this happens, is it using a lot of memory or cpu?

**Comment by andersevenrud** [comment]
@cseickel 

> Have you checked the process stats when this happens, is it using a lot of memory or cpu?

I noticed that memory usage was at around 7G for my neovim process, which is quite unusual, but as for CPU it looked normal (but I probably saw over all cores, and not the thread).

Sorry, but I don't have anything concrete at this moment. When this happened today I was under pressure with some work stuff, so in frustration I just restarted neovim. I will do some better profiling next time. All I know is that this only happens when I've used neo-tree in my session (and especially when left open).

**Comment by andersevenrud** [comment]
Okay, it just happened again. Memory usage started climbing and when I hit around 7G as before things started to lag like crazy. Watched it go all the way up to 9G, then I closed all open neo-tree's, and it went down to 150M and everything became responsive again.

Any tips for digging deeper into this ?

**Comment by rlch** [comment]
Yep, it seems as though my memory usage just keeps climbing; albeit at a **much** slower rate than when I initially made this issue (and I haven't touched my setup since aside from updating your plugin). So far I haven't run into any huge lag issues like @andersevenrud. Can also reproduce @andersevenrud 's point on freeing up the memory by closing the trees.

**Comment by cseickel** [comment]
Here are all the things that cause any background activity. If you turn all these off I would hope the problem goes away, then you can turn them back on incrementally to isolate what causes the issue:

```lua
local config = {
  enable_diagnostics = false,
  enable_git_status = false,
  enable_modified_markers = false, -- Show markers for files with unsaved changes.
  enable_refresh_on_write = false, -- Refresh the tree when a file is written. Only used if `use_libuv_file_watcher` is false.
  resize_timer_interval = -1, -- (default 500) in ms, needed for containers to redraw right aligned and faded content
                               -- set to -1 to disable the resize timer entirely
  --                           -- NOTE: this will speed up to 50 ms for 1 second following a resize

  filesystem = {
    use_libuv_file_watcher = false -- this is the default
  }
}
```

**Comment by cseickel** [comment]
Thanks @andersevenrud, I was able to recreate the issue with that test. I would bet that you `updatetime` or some other setting is creating more frequent updates for you, which makes the issue worse for you than it is for me. I don't notice any problem in my normal usage but holding enter for a while, or reversing that with backspace, does seem to leak memory.

I'll take a look and figure out where the leak is.

**Comment by cseickel** [comment]
I'm working on this but having a hard time finding it. I can tell that the issue is something to do with refreshing the filesystem source specifically. I haven't found any feature I can turn off which would fix the problem. Enabling `diagnostics` and holding down enter is a good test case because it will cause more refreshes than anything else.

**Comment by andersevenrud** [comment]
@cseickel Yeah, after a bit more testing this does indeed seem to be a general issue. On my desktop it's not really an issue (with `diagnostics` etc disabled), but my laptop really struggles after a while :(

Is it possible to make the tree close on opening a file? Sorry if this is obvious, but I checked the docs and found no mention. Considering how I use this plugin, having this option would basically be a workaround because I don't usually have a need to leave it open.

**Comment by cseickel** [comment]
@andersevenrud In addition to the `file_opened` event, you can also use the `float` and `current` position options. These will naturally go away as soon as you have opened a file. I'm a firm believer in using the `current` option when your goal is to just find a file to open without leaving the tree open.

**Comment by cseickel** [comment]
> I'm working on this but having a hard time finding it. I can tell that the issue is something to do with refreshing the filesystem source specifically. I haven't found any feature I can turn off which would fix the problem. Enabling `diagnostics` and holding down enter is a good test case because it will cause more refreshes than anything else.

☝️ That was entirely wrong. The reason I could not find the source of the leak was because it had nothing to do with Neo-tree. I finally just uninstalled Neo-tree and repeated the same test and found the same amount of memory leaked. 

That does not mean that there is not another memory leak related to Neo-tree, but it does mean that I do not actually have a way to recreate it. I can confirm that this memory leak that I did find is related to some other plugin I have, because it did not happen when I tried the minimal init that I use for testing. In fact, nvim with Neotree open stayed at a constant **17 MB!**

It's possible that the "hold down enter key" method of recreating the leak was actually just the discovery of a second source of memory leaks for you as well. To isolate Neo-tree as the source of the leak, can you come up with a method of recreating it starting with this command:
```
nvim -u tests/mininit.lua --noplugins
```

That `tests/mininit.lua` file is the root of this repo.

**Comment by nyngwang** [comment]
@cseickel if you find who is the culprit that causes the memory leak please let me know too!

**Comment by cseickel** [comment]
@nyngwang It's treesitter that causes the leak in the "enter insert mode and hold down enter" scenario. I went through the exercise of removing everything and adding them back in one at a time, and this leak starts when I add treesitter back in. It doesn't grow the leak with quite the same curve as I had before (it's more linear now), but I am betting it will get there as I enable the other treesitter related plugins.

**Comment by cseickel** [comment]
@andersevenrud @rlch The treesitter leak I discovered only happens in the neovim 0.8 series but not on an earlier 0.7 release. I am going to work on nailing down exactly where that issue started, but in the meantime can you tell me if you are running nightlies? If you are, can you try using a stable version of neovim and telling me if your issue still happens?

**Comment by andersevenrud** [comment]
@cseickel 

> but in the meantime can you tell me if you are running nightlies? If you are, can you try using a stable version of neovim and telling me if your issue still happens?

Yup. I usually compile from source a couple of times per week. I'll try checking out the 0.7 release.

**Comment by rlch** [comment]
> @andersevenrud @rlch The treesitter leak I discovered only happens in the neovim 0.8 series but not on an earlier 0.7 release. I am going to work on nailing down exactly where that issue started, but in the meantime can you tell me if you are running nightlies? If you are, can you try using a stable version of neovim and telling me if your issue still happens?

Yep, I use nightly as well. From memory this issue persisted for me in 0.7 and 0.8 but I'll investigate properly.

**Comment by nyngwang** [comment]
FYI: It seems that the issue is gone recently. I kept my laptop open(auto-sleep disabled) 24hrs+ with neo-tree opened and nothing laggy when I came back to it.

**Comment by nyngwang** [comment]
@rlch @soifou @andersevenrud Does anyone still experience this bug?

**Comment by nhat-vo** [comment]
This seems to have been solved and inactive for a while, so I'll close it as solved.
